pyg_lib/csrc/ops/cpu/matmul_kernel.cpp

```diff
@@ -1,46 +1,412 @@
+        std::vector<scalar_t> alpha(group_count, 1);
+        std::vector<scalar_t> beta(group_count, 0);
+
+        std::vector<int> ms(group_count);
+        std::vector<int> ns(group_count);
+        std::vector<int> ks(group_count);
+        std::vector<int> ld_src0(group_count);
+        std::vector<int> ld_src1(group_count);
+        std::vector<int> ld_dst(group_count);
+        std::vector<int> group_sizes(group_count);
+        std::vector<scalar_t*> src0;
+        std::vector<scalar_t*> src1;
+        std::vector<scalar_t*> dst;
+
+        size_t group_idx = 0;
+        for (const auto group_kv : groups) {
+          int m;
+          int n;
+          int k;
+          std::tie(m, n, k) = group_kv.first;
+          const auto indices = group_kv.second;
+
+          ms[group_idx] = m;
+          ns[group_idx] = n;
+          ks[group_idx] = k;
+          ld_src0[group_idx] = k;
+          ld_src1[group_idx] = n;
+          ld_dst[group_idx] = n;
+          group_sizes[group_idx] = indices.size();
+          ++group_idx;
+
+          for (const auto tensor_idx : indices) {
+            src0.push_back(input[tensor_idx].data_ptr<scalar_t>());
+            src1.push_back(other[tensor_idx].data_ptr<scalar_t>());
+            dst.push_back(out[tensor_idx].data_ptr<scalar_t>());
+          }
+        }
+
+        auto src0_ptrs = const_cast<const scalar_t**>(src0.data());
+        auto src1_ptrs = const_cast<const scalar_t**>(src1.data());
+        auto dst_ptrs = dst.data();
+
+#if AT_MKL_SEQUENTIAL()
+        // unlikely to happen - requires Torch to be built from source with
+        // explicit flag denoting MKL sequential version
+        parallel_mkl_blas_gemm_batched(ms, ns, ks, alpha, src0_ptrs, ld_src0,
+                                       src1_ptrs, ld_src1, beta, dst_ptrs,
+                                       ld_dst, group_count, group_sizes);
+#else
+        mkl_blas_gemm_batched(ms.data(), ns.data(), ks.data(), alpha.data(),
+                              src0_ptrs, ld_src0.data(), src1_ptrs, ld_src1.data(),
+                              beta.data(), dst_ptrs, ld_dst.data(), group_count,
+                              group_sizes.data());
+#endif
+      });
 }
 
 std::vector<at::Tensor> grouped_matmul_kernel(const at::TensorList input,
                                               const at::TensorList other) {
-  std::vector<at::Tensor> out(input.size());
-  for (size_t i = 0; i < input.size(); ++i)
-    out[i] = input[i].new_empty({input[i].size(0), other[i].size(-1)});
+  const auto n_matrices = input.size();
 
-  grouped_matmul_out_kernel(input, other, out);
+  std::vector<at::Tensor> input_contig;
+  std::vector<at::Tensor> other_contig;
+  std::vector<at::Tensor> out;
+
+  input_contig.reserve(n_matrices);
+  other_contig.reserve(n_matrices);
+  out.reserve(n_matrices);
+
+  for (size_t i = 0; i < n_matrices; ++i) {
+    input_contig.emplace_back(input[i].contiguous());
```

Comment:
this is OK if we expect the `input` and `other` are given in the contiguous format already.
if not, say `other` is transposed, we need to change the logic here.